### PR TITLE
Modified serializer and views

### PR DESCRIPTION
**Pull request by TkTioNG**
Update serializers and views  to pass the unit tests written

Remove float casting in get_total_price in utils

Cast Decimal object into string object (same as the JSON format)

### CODE REVIEW

**Comment by TkTioNG** [comment]
Add validation on ProductCapacitySerializer for remaining_capacities field
Refactor validation for SalesSerializer
Refactor utils function for get_restock_total_price

File: inventory/serializers.py [hunk 1]
```diff
@@ -86,27 +88,21 @@ class ProductCapacitySerializer(serializers.ModelSerializer):
     remaining_capacities = serializers.SerializerMethodField()
 
     def get_remaining_capacities(self, obj):
-        data = []
-        for product in obj.products.all():
-            material_quantities_need = MaterialQuantity.objects.filter(
-                product=product
+        return get_product_remaining_capacities(obj)
```
Comment (AbodSinan) [inline inventory/serializers.py:91]: usually functions like `get_product_remaining_capacities()` are called "service calls" and are stored in a `services` folder in the project, where the business logic is usually processed. Utils are more for utility functions that perform simple tasks like conversion or general calculation/rounding.

File: inventory/serializers.py [hunk 2]
```diff
@@ -118,73 +114,63 @@ class SalesSerializer(serializers.Serializer):
     sale = serializers.SerializerMethodField()
 
     def get_sale(self, obj):
-        data = []
-        for product in obj.products.all():
-            material_quantities_need = MaterialQuantity.objects.filter(
-                product=product
+        return get_product_remaining_capacities(obj)
+
+    def is_valid(self, raise_exception=False):
```
Comment (AbodSinan) [inline inventory/serializers.py:119]: okay yea this is easier to read

File: inventory/serializers.py [hunk 3]
```diff
@@ -118,73 +114,63 @@ class SalesSerializer(serializers.Serializer):
...
+        super().is_valid(raise_exception)
+
+        if "sale" not in self.initial_data:
+            raise ValidationError(detail="Missing Information in sales")
+
+        # if sale is post, there must be products exist in the data
+        sale = self.initial_data.get("sale")
+        if not hasattr(sale, "__len__") or len(sale) == 0:
```
Comment (AbodSinan) [inline inventory/serializers.py:127]: i wouldn't check against the metadata of an object, it's a bit volatile. You could use `bool(dict)` to check whether a dictionary is empty, or if you want to check a list `len(list)` would suffice.

TLDR: try not to use hasattr() unless absolutely necessary, there are better ways to do it in most cases.

File: inventory/serializers.py [hunk 4]
```diff
@@ -118,73 +114,63 @@ class SalesSerializer(serializers.Serializer):
...
-            raise ValidationError(self._errors)
+        self._validated_data["sale"] = self.initial_data["sale"]
 
         return not bool(self._errors)
 
+    def validate_material_sufficiency(self, materials_needed, sold_quantity):
+        # Check the sufficiency of the material, if not enough, raise error
+        for material in materials_needed:
+            material_quantity = material.quantity
+            material_stock_obj = self.instance.material_stocks.get(
```
Comment (AbodSinan) [inline inventory/serializers.py:160]: Usually when you do object.get() you should put it in a try clause and catch a DoesNotExist error in case no instance is found.

File: inventory/tests/viewsets/test_material_viewset.py [hunk 5]
```diff
@@ -40,9 +41,12 @@ def _get_expected_params(self, objects, many=False):
         if many:
             expected_params = []
             for obj in objects:
-                expected_params.append(
-                    get_model_obj_property(Material, obj)
+                material_obj = get_model_obj_property(Material, obj)
```
Comment (AbodSinan) [inline inventory/tests/viewsets/test_material_viewset.py:44]: I really like the get_model_obj_property function, would help alot in unit testing

File: inventory/tests/viewsets/test_sales_viewset.py [hunk 6]
```diff
@@ -83,7 +84,7 @@ def test_sold_product_valid(self):
         self.assertEqual(response.status_code, 200)
 
         # Verify the sales content
-        self.assertEqual(response.json(), self.data)
+        self.assertEqual(response.json(), self._get_expected_params())
```
Comment (AbodSinan) [inline inventory/tests/viewsets/test_sales_viewset.py:87]: it's better to explicitly set the expected params in the unit test, because in this case if the get_product_remaining_capacities function isn't working properly the test will still pass because it is used in the processing the request(in serializer/view) as well as in the unit test.

File: inventory/utils.py [hunk 7]
```diff
@@ -1,18 +1,14 @@
-from inventory.models import Material, MaterialStock
+from inventory.models import Material, MaterialStock, MaterialQuantity, Store
```
Comment (AbodSinan) [inline inventory/utils.py:2]: The same note from a comment above, these are business logic functions that should be under Services rather than Utils
